File: com/white/AddTwoNumbers.java
package com.white;

public class AddTwoNumbers {

    // problem 2

    public static ListNode addTwoNumbers(ListNode l1, ListNode l2) {

        ListNode result = new ListNode(0);
        ListNode x1 = l1;
        ListNode x2 = l2;
        ListNode temp = result;
        int carry = 0;

        while(x1!=null || x2!=null || carry==1) {
            int sum = 0;
            if (x1!=null) {
                sum += x1.val;
                x1 = x1.next==null ? null : x1.next;
            }
            if (x2!=null) {
                sum += x2.val;
                x2 = x2.next==null ? null : x2.next;
            }
            temp.next = new ListNode(sum+carry);
            temp = temp.next;
            carry = temp.val / 10;
            temp.val = temp.val % 10;
        }

        return result.next;
    }
}

File: LeetcodeCore/AddTwoNumbers.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetcodeCore
{
    public class AddTwoNumbersSolution
    {
        // 2. Add Two Numbers
        public ListNode AddTwoNumbers(ListNode l1, ListNode l2)
        {
            var a = l1.val;
            var b = l2.val;
            var c = 0;
            var dummy = new ListNode(0);
            var curr = dummy;

            while (a != 0 || b != 0 || c != 0 || l1 != null || l2 != null)
            {
                var sum = a + b + c;
                var mod = sum % 10;
                curr.next = new ListNode(mod);
                curr = curr.next;
                c = sum / 10;
                l1 = l1?.next;
                l2 = l2?.next;
                a = l1 == null ? 0 : l1.val;
                b = l2 == null ? 0 : l2.val;
            }

            return dummy.next;
        }
    }
}
